_posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md

```diff
@@ -0,0 +1,196 @@
+We created another queue for our batch messages, and when consuming them from the original queue – we just re-queued them to another. We were using [RabbitMqBundle](https://github.com/php-amqplib/RabbitMqBundle) for consuming messages from the queue and Symfony Messenger inside application.
+
+We had to change routing for the specific message in order to put it on the different queue:
+
+{% highlight yaml %}
+#config/packages/messenger.yaml
+framework:
+    messenger:
+        transports:
+            batch_messages_amqp: '%env(BATCH_MESSAGES_TRANSPORT_DSN)%/advert-served-to-user
+        ...
+        routing:
+            App\Message\Command\AdvertServedToUser: batch_messages_amqp
+
+{% endhighlight %}
+
+The next step was to get messages from the new queue in a batch of 50 or more (we ended up getting 5000 of them in one batch). In order to do that we had to create consumer class:
+
+{% highlight php %}
+<?php
+
+declare(strict_types=1);
+
+namespace App\Consumer;
+
+use App\Message\Command\Advert\AdvertServedToUser;
+use App\Message\Command\Advert\RunAdvertServedToUserBatch;
+use OldSound\RabbitMqBundle\RabbitMq\BatchConsumerInterface;
+use PhpAmqpLib\Message\AMQPMessage;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Messenger\MessageBusInterface;
+
+class AdvertServedToUserToUserBatchConsumer implements BatchConsumerInterface
+{
+    private $commandBus;
+    private $logger;
+
+    public function __construct(MessageBusInterface $commandBus, LoggerInterface $logger)
+    {
+        $this->commandBus = $commandBus;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @param AMQPMessage[] $messages
+     *
+     * @return array|bool
+     */
+    public function batchExecute(array $messages)
+    {
+        try {
+            $command = new RunAdvertServedToUserBatch($messages);
+            $this->commandBus->dispatch($command);
+            return true;
+        } catch (\Throwable $e) {
+            $this->logger->log('error', $e->getMessage(), ['exception' => $e]);
+            return false;
+        } 
+    }
+}
+{% endhighlight %}
+
+registered class as a service:
+
+{% highlight yaml %}
+#config/services.yaml
+services:
+    ...
+    app.consumer.batch:
+        class: App\Consumer\AdvertServedToUserBatchConsumer
+        arguments: ["@messenger.bus.command", '@logger']
+{% endhighlight %}
+
+and use our consumer as callback for batch RabbitMQ
+
+{% highlight yaml %}
+#config/packages/old_sound_rabbit_mq.yaml
+old_sound_rabbit_mq:
+    connections:
+        default:
+            ...
+        advert_served_to_user:
+            url: '%env(BATCH_MESSAGES_TRANSPORT_DSN)%'
+            lazy:     true
+            connection_timeout: 3
+            read_write_timeout: 3
+            # requires php-amqplib v2.4.1+ and PHP5.4+
+            keepalive: false
+            # requires php-amqplib v2.4.1+
+            heartbeat: 0
```

Comment:
I would remove the parameters here that is not needed

Reply:
Which ones?

Reply:
lazy, connection_timeout, read_write_timeout, keepalive, heartbeat